src/webauthn/proto/raw_message.rs

```diff
@@ -45,14 +58,525 @@ pub struct Packed {
 }
 
 #[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
 pub struct TPM {
+    pub ver: serde_cbor::Value,
     pub alg: i64,
     #[serde(with = "serde_bytes")]
     pub sig: Vec<u8>,
     #[serde(skip_serializing_if = "Option::is_none")]
     pub x5c: Option<serde_cbor::Value>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub ecdaa_key_id: Option<serde_cbor::Value>,
+    #[serde(with = "serde_bytes")]
+    pub cert_info: Vec<u8>,
+    #[serde(with = "serde_bytes")]
+    pub pub_area: Vec<u8>,
+}
+
+impl TPM {
+    pub fn verify_structure(&self) -> Result<CertInfo, Error> {
+        if CoseAlgorithmIdentifier::from(self.alg) == CoseAlgorithmIdentifier::NotSupported {
+            return Err(Error::TpmError(TpmError::AlgorithmNotSupported));
+        }
+
+        if let Value::Text(ver) = &self.ver {
+            if ver != "2.0" {
+                return Err(Error::TpmError(TpmError::AttestationVersionNotSupported));
+            }
+        } else {
+            return Err(Error::TpmError(TpmError::AttestationVersionNotSupported));
+        }
+
+        if self.x5c.is_none() {
+            return Err(Error::TpmError(TpmError::CertificateMissing));
+        }
+
+        let cert_info = CertInfo::from_buf(self.cert_info.as_slice())?;
```

Comment:
```suggestion
        let cert_info = CertInfo::from_slice(self.cert_info.as_slice())?;
```
See change in `impl CertInfo`